### ISSUE TITLE
Relation where() with Closure expects wrong argument because of Builder mixin

### ISSUE THREAD

**Issue by rudiedirkx** [nunomaduro/larastan#2078]
- Larastan Version: v2.9.9
- Laravel Version: v11.23.5

### Description

Sorry for the shit title 😆 

I want to fetch some eloquent objects, starting from a relationship query. One of the extra conditions (beside the relation condition itself) is a `where()` with a Closure (to do an `orWhere()` in). The `where()` Closure argument is a subquery builder of type EloquentBuilder. But because `where()` is called on a `HasMany` query, not an `EloquentBuilder` query, Larastan/Laravel wants the builder argument to be a `HasMany`, which is very wrong. I think it's because HasMany pipes all methods to EloquentBuilder, but Larastan doesn't know that when it evaluates `where()`'s Closure.

Sorry for the shit description! This is hard. The example should be pretty obvious:

### Laravel code where the issue was found

```php
$query = $organization->sites()
	->where(function(EloquentBuilder $query) use ($search) {
		$query->where('name', 'like', "%$search%");
		$query->orWhere('url', 'like', "%$search%");
	});
```

Phpstan throws a giant error about the Closure (really about its argument `EloquentBuilder $query`):

    Parameter #1 $column of method Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Site,App\Models\Organization>::where() expects array<int|string, mixed>|(Closure(Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Site, App\Models\Organization>): mixed)|Illuminate\Contracts\Database\Query\Expression|string, Closure(Illuminate\Database\Eloquent\Builder): void given.

Not very readable, but one of the accepted types is

    (Closure(Illuminate\Database\Eloquent\Relations\HasMany<App\Models\Site, App\Models\Organization>): mixed)

which is **almost** right! But not quite, because it says `Relations\HasMany` and that's not how `where()` works.

----

There is a very simple workaround (which might actually be the solution instead of a workaround):

```php
$query = $organization->sites()
	->getQuery()
	->where(function(EloquentBuilder $query) use ($search) {
\PHPStan\dumpType($query);
		$query->where('name', 'like', "%$search%");
		$query->orWhere('url', 'like', "%$search%");
	});
```

The `getQuery()` makes the `HasMany` relationship query a real `EloquentBuilder` query, and then Larastan understands what `$query` is and should be:

    Dumped type: Illuminate\Database\Eloquent\Builder<App\Models\Site>

----

I kinda like the workaround, because it makes the new query explicit. It's not a relationship anymore, but a new query. Buuuut it might still be a bug. You decide.

**Comment by calebdw** [comment]
The following works just fine in the framework type tests:

```php
$user->roles()->where(function ($query) {
    assertType('Illuminate\Database\Eloquent\Builder<Illuminate\Types\Relations\Role>', $query);
});
```

**Comment by rudiedirkx** [comment]
Odd... Maybe because it's not typed at all?

**Comment by szepeviktor** [comment]
Yes!! Typed closures are beneficial for everything. (except for Larastan)

**Comment by rudiedirkx** [comment]
Are we okay with that? I as a human kind of like types, because I can see what I'm working with. I want to keep Larastan happy of course, but I also want to be happy!

But I'm happy with the `getQuery()` solution, so I'm happy anyway.

**Comment by canvural** [comment]
This was a tricky one to solve. But it is fixed with df279123

Thanks.
